### PR TITLE
Issue 20071

### LINKED ISSUE #20071: System Template
We need to implement **Templates as a File**, those templates will live under `application/templates/` and have the following files:
* **Properties.vtl**:
```
$dotJSON.put("title", "System Template")
$dotJSON.put("friendly_name", "The System Template...")

## This properties only if is a Design Template
$dotJSON.put("theme", "system_theme")
```

* If it is an Advanced Template **Body.vtl** e.g
```
#parseContainer('/application/containers/banner/','1615825205295')
   
```
* If it is a Design Template **layout.json** e.g (this is the system template)
```
{
      "body":{
         "rows":[
            {
               "styleClass":"",
               "columns":[
                  {
                     "styleClass":"",
                     "leftOffset":1,
                     "width":12,
                     "containers":[
                        {
                           "identifier":"//default/application/containers/system/",
                           "uuid":"1"
                        }
                     ]
                  }
               ]
            }
         ]
      },
      "header":true,
      "footer":true,
      "sidebar":null
}
```

System Template will be shipped in the empty starter and will be a Design Template as a file.

When a new page is created and there is no `Default Template` available, the page should default to `SYSTEM_TEMPLATE`. 

The SYSTEM_TEMPLATE should show the header/footer and have 1 full with a row with the system_container in it.

We need to test PP of Templates as a File.

Design Template as a File should support layouts?? **YES**

When creating a page that uses a Template as a File, the template Id should be saved as a Path.

### PR DESCRIPTION
**Pull request by erickgonzalez**
This PR is meant to achieve the creation of file asset templates, these templates will live under **application/templates/**. Templates under it could be Design ones or Advanced, this will be diff by the files the user uploads: **layout.json** for the Design, a **body.vtl** for the Advanced; regardless of the type of the Template **properties.vtl** must be uploaded with the metadata of the template.

- New empty starter was created with the system template.
- Tests.
- Changes for PP File Asset Templates, to pull those as a dependency when PP pages.
- Changes in the TemplateLoader, so it can render pages with file asset templates since its identifier is a path.
- A job was created to update the template of a page when a rename of the site is executed.
- A listener was created to listen to all the possible file changes in the templates folder.

### CODE REVIEW

File: dotCMS/src/main/java/com/dotmarketing/portlets/templates/business/FileAssetTemplateUtil.java [hunk 1]
```diff
@@ -0,0 +1,497 @@
...
+                APILocator.getHostAPI().resolveHostName(hostname, APILocator.systemUser(), false);
+    }
+
+    final List<String> pageModePrefixList = Stream.of(PageMode.values())
+            .map(pageMode -> String.format("/%s/", pageMode.name()))
+            .collect(Collectors.toList());
+
+
+    //demo.dotcms.com/application/templates/test/
+    public String getHostName(final String path) {
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotmarketing/portlets/templates/business/FileAssetTemplateUtil.java:132]: I think this logic is kinda similar to the container, probably would be a good idea to unified the logic in just one class and pass more parameters (for container or templates)
In that case, if we found an error probably can fix just once

File: dotCMS/src/main/java/com/dotmarketing/portlets/templates/business/FileAssetTemplateUtil.java [hunk 2]
```diff
@@ -0,0 +1,497 @@
...
+                if (tmp.startsWith(prefix)) {
+                    tmp = tmp.substring(prefix.length());
+                    break;
+                }
+            }
+
+            tmp = tmp.replaceAll(HOST_INDICATOR, StringPool.BLANK);
+            tmp = tmp.substring(0, tmp.indexOf(TEMPLATE_FOLDER_PATH));
+            final String finalString = tmp;
+            Logger.debug(FileAssetTemplateUtil.class,
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotmarketing/portlets/templates/business/FileAssetTemplateUtil.java:146]: I am not 100% sure if the String.format may support a null. I would do something like

final String finalString = UtilMethods.isSet(tmp) ? tmp : StringPool.NULL

File: dotCMS/src/main/java/com/dotmarketing/portlets/templates/business/FileAssetTemplateUtil.java [hunk 3]
```diff
@@ -0,0 +1,497 @@
...
+                host = this.getHostFromHostname(hostname);
+            }
+        } catch (DotDataException | DotSecurityException e) {
+            host = null;
+        }
+
+        if (null == host) {
+
+            try {
+                host = WebAPILocator.getHostWebAPI()
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotmarketing/portlets/templates/business/FileAssetTemplateUtil.java:174]: the current request could be null, should catch NPE exception

File: dotCMS/src/main/java/com/dotmarketing/portlets/templates/business/FileAssetTemplateUtil.java [hunk 4]
```diff
@@ -0,0 +1,497 @@
...
+     * Wraps the file asset into the dotParse directive, this is helpful in order to fetch lazy on runtime the fileasset and also, to add multi lang capabilities
+     * @param fileAsset {@link FileAsset}
+     * @return String
+     */
+    public String wrapIntoDotParseDirective (final FileAsset fileAsset) {
+
+        try {
+
+            final Host host = APILocator.getHostAPI().find(fileAsset.getHost(), APILocator.systemUser(), false);
+            String  dotParseFilePath = "//" + host.getHostname()  + fileAsset.getPath() + fileAsset.getFileName()  ;
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotmarketing/portlets/templates/business/FileAssetTemplateUtil.java:301]: set to final

File: dotCMS/src/main/java/com/dotmarketing/portlets/templates/business/FileAssetTemplateUtil.java [hunk 5]
```diff
@@ -0,0 +1,497 @@
...
+            Logger.debug(FileAssetTemplateUtil.class,()->"dotParse directive file uri: " + dotParseFilePath);
+            return String.format("#dotParse(\"%s\")",dotParseFilePath);
+
+        } catch (DotSecurityException | DotDataException  e) {
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotmarketing/portlets/templates/business/FileAssetTemplateUtil.java:305]: Log the error

File: dotCMS/src/main/java/com/dotmarketing/portlets/templates/business/TemplateAPIImpl.java [hunk 6]
```diff
@@ -604,6 +632,17 @@ public boolean delete(final Template template, final User user, final boolean re
 	}
 
 	public Template findWorkingTemplate(final String id, final User user, final boolean respectFrontendRoles) throws DotDataException, DotSecurityException {
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotmarketing/portlets/templates/business/TemplateAPIImpl.java:634]: need CloseDBIfOpened Annotation

File: dotCMS/src/main/java/com/dotmarketing/portlets/templates/business/TemplateAPIImpl.java [hunk 7]
```diff
@@ -631,6 +670,17 @@ public Template find(final String inode, final User user, final boolean respectF
 	}
 
 	public Template findLiveTemplate(final String id, final User user, final boolean respectFrontendRoles) throws DotDataException, DotSecurityException {
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotmarketing/portlets/templates/business/TemplateAPIImpl.java:672]: Add closeDbIfOpened annotation

**Review by jdotcms — APPROVED** [review]
Very good job

File: dotCMS/build.gradle [hunk 8]
```diff
@@ -181,10 +181,10 @@ dependencies {
     /**** And now the libs we pull in from internal company sources - libs stored in ./plugins, ./bin, ./libs, the starter site, etc. ****/
     compile fileTree("src/main/plugins/com.dotcms.config/build/jar").include('plugin-com.dotcms.config.jar')
 
-    starter group: 'com.dotcms', name: 'starter', version: 'empty_20210315', ext: 'zip'
+    starter group: 'com.dotcms', name: 'starter', version: 'empty_20210405', ext: 'zip'
```
Comment (nollymar) [inline dotCMS/build.gradle:184]: A new starter version will be required when `release-21.04` gets merged into `master`

File: dotCMS/src/integration-test/java/com/dotcms/datagen/TemplateAsFileDataGen.java [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+            FileUtil.write(fileMetaData, isDesignTemplate ? metaDataDesign : metaDataBasic);
+            final Contentlet templateMetaDataFileAsset = new FileAssetDataGen(templateFolder,fileMetaData)
+                    .host(host).setProperty("title",Constants.TEMPLATE_META_INFO_FILE_NAME)
+                    .setProperty("fileName",Constants.TEMPLATE_META_INFO_FILE_NAME)
+                    .nextPersisted();
+            metaData = APILocator.getFileAssetAPI().fromContentlet(templateMetaDataFileAsset);
+
+            return toFileAssetTemplate(object,templateFolder,metaData,templateCode);
+        } catch (DotDataException | DotSecurityException | IOException e) {
+            throw new RuntimeException(e);
```
Comment (nollymar) [inline dotCMS/src/integration-test/java/com/dotcms/datagen/TemplateAsFileDataGen.java:99]: Should we need a log here with more info? maybe a WarnDebug

File: dotCMS/src/main/java/com/dotcms/publisher/util/DependencyManager.java [hunk 10]
```diff
@@ -887,6 +904,28 @@ private void processTemplateDependencies(final String id) {
...
+			final List<Folder> subFolders = APILocator.getFolderAPI()
+					.findSubFolders(rootFolder, user, false);
+
+			final Set<Folder> dependenciesFolders = new HashSet<>();
+			dependenciesFolders.add(rootFolder);
+			dependenciesFolders.addAll(subFolders);
+
+			return dependenciesFolders;
+		}catch (DotSecurityException | DotDataException e) {
+			Logger.error(DependencyManager.class, e);
```
Comment (nollymar) [inline dotCMS/src/main/java/com/dotcms/publisher/util/DependencyManager.java:924]: I would add a message including the fileAssetTemplate that failed

File: dotCMS/src/main/java/com/dotmarketing/portlets/contentlet/business/web/UpdatePageTemplatePathJob.java [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+import org.quartz.JobExecutionException;
+import org.quartz.Scheduler;
+import org.quartz.SchedulerException;
+import org.quartz.SimpleTrigger;
+
+/**
+ * Should be run after a change in a host's name, it update all the file asset templates of pages
+ * to use the new host's name.
+ */
+public class UpdatePageTemplatePathJob extends DotStatefulJob {
```
Comment (nollymar) [inline dotCMS/src/main/java/com/dotmarketing/portlets/contentlet/business/web/UpdatePageTemplatePathJob.java:31]: Is it possible to implement a test for the logic in the `run` method?

File: dotCMS/src/main/java/com/dotmarketing/portlets/templates/business/FileAssetTemplateUtil.java [hunk 12]
```diff
@@ -0,0 +1,409 @@
...
+        try {
+            final FileAsset asset = APILocator.getFileAssetAPI().fromContentlet(contentlet);
+            final Folder parentFolder = APILocator.getFolderAPI()
+                    .find(asset.getFolder(), APILocator.systemUser(), false);
+
+            return (TEMPLATE_META_INFO.equals(asset.getFileName()) && UtilMethods
+                    .isSet(parentFolder.getPath()) && parentFolder.getPath()
+                    .startsWith(TEMPLATE_FOLDER_PATH));
+        } catch (Exception e) {
+            Logger.error(FileAssetTemplateUtil.class, "Unable to determine if contentlet is a fileAssetTemplate ", e);
```
Comment (nollymar) [inline dotCMS/src/main/java/com/dotmarketing/portlets/templates/business/FileAssetTemplateUtil.java:370]: Logging the contentlet's inode might be helpful for support

File: dotCMS/src/integration-test/java/com/dotcms/rendering/velocity/services/HTMLPageAssetRenderedTest.java [hunk 13]
```diff
@@ -1853,6 +1855,77 @@ public void pageWithSidebarShouldRender() throws Exception{
         assertTrue(firstIndex != -1 && secondIndex != -1 && firstIndex != secondIndex && thirdIndex == -1);
     }
 
+    /**
+     * Method to test: {@link HTMLPageAssetRenderedAPI#getPageHtml(PageContext, HttpServletRequest, HttpServletResponse)}
+     * When: A page has a template that is a file template advanced
+     * Should: render the page
+     */
+    @Test
```
Comment (freddyucv) [inline dotCMS/src/integration-test/java/com/dotcms/rendering/velocity/services/HTMLPageAssetRenderedTest.java:1863]: I think this test is with a not advanced template

File: dotCMS/src/main/java/com/dotcms/publisher/util/DependencyManager.java [hunk 14]
```diff
@@ -878,6 +889,12 @@ private void processTemplateDependencies(final String id) {
 					}
 				}
 			}
+
+			if(workingTemplate instanceof FileAssetTemplate){
+				//Process FileAssetTemplate
+				final Set<Folder> folders = this.collectFileAssetTemplateDependencies(FileAssetTemplate.class.cast(workingTemplate));
+				setFolderListDependencies(folders);
```
Comment (freddyucv) [inline dotCMS/src/main/java/com/dotcms/publisher/util/DependencyManager.java:896]: really you not need to call  setFolderListDependencies here, just add the folders into the dependencyProcesor
```suggestion
				dependencyProcessor.putAll(folders, AssetTypes.FOLDER);
```

really the putAll method not exists but it is easy to implement

The dependencyProcesor call the setFolderListDependencies later

File: dotCMS/src/main/java/com/dotcms/rest/api/v1/template/TemplateHelper.java [hunk 15]
```diff
@@ -66,7 +68,8 @@ public TemplateView toTemplateView(final Template template, final User user) {
                 .name(template.getName())
                 .friendlyName(template.getFriendlyName())
                 .title(template.getTitle())
-                .identifier(template.getIdentifier())
+                .identifier(FileAssetTemplateUtil.getInstance().isFileAssetTemplate(template)?
```
Comment (freddyucv) [inline dotCMS/src/main/java/com/dotcms/rest/api/v1/template/TemplateHelper.java:71]: I dont know if it is good idea do this here, maybe it is better in a TemplateAPI or IdentifierAPI, maybe a method getTemplateIdentifier

File: dotCMS/src/main/java/com/dotmarketing/factories/PublishFactory.java [hunk 16]
```diff
@@ -357,7 +359,17 @@ public static boolean publishHTMLPage(IHTMLPage htmlPage, List relatedNotPublish
...
+					//inode of the template is the inode of the properties.vtl
+					//we need to obtain the folder so publish all the files related to template as files
+					final Contentlet propertiesVTL = APILocator.getContentletAPI().find(((FileAssetTemplate) asset).getInode(),user,false);
+					final Identifier idPropertiesVTL = APILocator.getIdentifierAPI().find(propertiesVTL.getIdentifier());
+					final Folder templateFolder = APILocator.getFolderAPI().findFolderByPath(idPropertiesVTL.getParentPath(),
+							idPropertiesVTL.getHostId(),user,respectFrontendRoles);
+					publishAsset(templateFolder,user,respectFrontendRoles,false);
+				} else {
+					publishAsset((Template) asset, user, respectFrontendRoles, false);
+				}
```
Comment (freddyucv) [inline dotCMS/src/main/java/com/dotmarketing/factories/PublishFactory.java:372]: maybe this code should go into a API?

**Review by freddyucv — APPROVED** [review]

**Review by nollymar — APPROVED** [review]

**Pull request merged**